File: 2022/postgresql_insert_performance/go/main.go
package main

import (
	"context"
	"fmt"
	"log"
	"math/rand"
	"os"
	"strings"
	"time"

	"github.com/google/uuid"
)

const CONCURRENCY = 100
const EXECUTIONS = 30_000
const RUNS = 7

type Event struct {
	ID         uuid.UUID
	Type       string
	Timestamp  time.Time
	ReceivedAt time.Time
	Payload    Payload
}

type Payload struct {
	Something      string
	SomethingElse  []byte
	SomethingElse2 []uint32
}

type KeyValueEvent struct {
	Key   uuid.UUID
	Value []byte
}

func main() {
	ctx := context.Background()
	databaseUrl := os.Getenv("DATABASE_URL")
	if databaseUrl == "" {
		log.Fatal("env var DATABASE_URL is missing")
	}

	pool, err := dbConnect(ctx, databaseUrl)
	if err != nil {
		log.Fatal(err)
	}
	defer pool.Close()

	err = dbSetup(ctx, pool)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("DB Setup")

	// normalized
	fmt.Println("Normalized")
	normalizedResults := make([]time.Duration, RUNS)
	for i := 0; i < RUNS; i++ {
		dbCleanTable(ctx, pool, "normalized")
		start := time.Now()
		errInsert := insertNormalized(ctx, pool)
		end := time.Now()
		normalizedResults[i] = end.Sub(start)
		if errInsert != nil {
			log.Fatal(errInsert)
		}
		fmt.Printf("%d,", i)
	}
	fmt.Println("")
	fmt.Printf("    results: %v", normalizedResults)
	normalizedMean := durationMean(normalizedResults)
	fmt.Printf("    mean: %v", normalizedMean)

	fmt.Println("\n------------------------------------------\n")

	// normalized
	fmt.Println("Key Value")
	keyValueResults := make([]time.Duration, RUNS)
	for i := 0; i < RUNS; i++ {
		dbCleanTable(ctx, pool, "key_value")
		start := time.Now()
		errInsert := insertKeyValue(ctx, pool)
		end := time.Now()
		keyValueResults[i] = end.Sub(start)
		if errInsert != nil {
			log.Fatal(errInsert)
		}
	}
	fmt.Printf("    results: %v", keyValueResults)
	keyValueMean := durationMean(keyValueResults)
	fmt.Printf("    mean: %v", keyValueMean)

	fmt.Println("\n------------------------------------------\n")

	// time series
	fmt.Println("Timeseries")
	timeseriesResults := make([]time.Duration, RUNS)
	for i := 0; i < RUNS; i++ {
		dbCleanTable(ctx, pool, "timeseries")
		start := time.Now()
		errInsert := insertTimeSeries(ctx, pool)
		end := time.Now()
		timeseriesResults[i] = end.Sub(start)
		if errInsert != nil {
			log.Fatal(errInsert)
		}
	}
	fmt.Printf("    results: %v", timeseriesResults)
	timeseriesMean := durationMean(timeseriesResults)
	fmt.Printf("    mean: %v", timeseriesMean)
}

func durationMean(results []time.Duration) time.Duration {
	var ret time.Duration

	for _, result := range results {
		ret += result
	}

	return ret / time.Duration(len(results))
}

func generateEvent() Event {
	now := time.Now().UTC()

	payload := Payload{
		Something:      RandString(500),
		SomethingElse:  make([]byte, 500),
		SomethingElse2: make([]uint32, 500),
	}

	return Event{
		ID:         uuid.New(),
		Type:       "some.event.type",
		Timestamp:  now,
		ReceivedAt: now,
		Payload:    payload,
	}
}

const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
const (
	letterIdxBits = 6                    // 6 bits to represent a letter index
	letterIdxMask = 1<<letterIdxBits - 1 // All 1-bits, as many as letterIdxBits
	letterIdxMax  = 63 / letterIdxBits   // # of letter indices fitting in 63 bits
)

func RandString(n int) string {
	var src = rand.NewSource(time.Now().UnixNano())
	sb := strings.Builder{}
	sb.Grow(n)
	// A src.Int63() generates 63 random bits, enough for letterIdxMax characters!
	for i, cache, remain := n-1, src.Int63(), letterIdxMax; i >= 0; {
		if remain == 0 {
			cache, remain = src.Int63(), letterIdxMax
		}
		if idx := int(cache & letterIdxMask); idx < len(letterBytes) {
			sb.WriteByte(letterBytes[idx])
			i--
		}
		cache >>= letterIdxBits
		remain--
	}

	return sb.String()
}

File: 2022/postgresql_insert_performance/src/main.rs
use chrono::{DateTime, Utc};
use futures::{stream, StreamExt};
use serde::{Deserialize, Serialize};
use sqlx::{postgres::PgPoolOptions, types::Json, Executor, Pool, Postgres};
use std::time::Duration;
use uuid::Uuid;

type DB = Pool<Postgres>;
const CONCURRENCY: u32 = 100;
const EXECUTIONS: u64 = 100_000;

#[derive(sqlx::FromRow, Serialize, Debug, Clone, Deserialize)]
struct Event {
    id: uuid::Uuid,
    r#type: String,
    timestamp: DateTime<Utc>,
    received_at: DateTime<Utc>,
    payload: Json<Payload>,
}

#[derive(sqlx::FromRow, Serialize, Debug, Clone, Deserialize)]
struct KeyValueEvent {
    key: uuid::Uuid,
    value: Vec<u8>,
}

#[derive(Serialize, Debug, Clone, Deserialize)]
struct Payload {
    something: String,
    something_else: Vec<u8>,
    something_else2: Vec<u32>,
}

#[tokio::main]
async fn main() -> Result<(), anyhow::Error> {
    let database_url = std::env::var("DATABASE_URL").expect("DATABASE_URL env is missing");

    Ok(())
}

async fn db_connect(database_url: &str) -> Result<DB, anyhow::Error> {
    let pool = PgPoolOptions::new()
        .max_connections(CONCURRENCY)
        .connect_timeout(Duration::from_secs(10))
        .connect(&database_url)
        .await?;
    Ok(pool)
}

fn generate_event() -> Event {
    let now = Utc::now();
    let payload = Payload {
        something: String::new(),
        something_else: Vec::new(),
        something_else2: Vec::new(),
    };

    Event {
        id: Uuid::new_v4(),
        r#type: "some.event.type".to_string(),
        timestamp: now,
        received_at: now,
        payload: Json(payload),
    }
}

async fn db_setup(db: &DB) -> Result<(), anyhow::Error> {
    db.execute(
        "
    CREATE TABLE IF NOT EXISTS normalized (
        id UUID PRIMARY KEY,
        type TEXT NOT NULL,
        timestamp TIMESTAMP WITH TIME ZONE NOT NULL,
        received_at TIMESTAMP WITH TIME ZONE NOT NULL,
        payload JSONB NOT NULL
    );

    CREATE TABLE IF NOT EXISTS key_value (
        key UUID PRIMARY KEY,
        value BYTEA NOT NULL
    );

    CREATE TABLE IF NOT EXISTS key_value_compressed (
        key UUID PRIMARY KEY,
        value BYTEA NOT NULL
    );
    ",
    )
    .await?;
    Ok(())
}

async fn clean_table(db: &DB, table: &str) {
    let query_delete = "DELETE FROM ".to_string() + table;
    let query_vacuum = "VACUUM FULL ".to_string() + table;

    sqlx::query(&query_delete)
        .execute(db)
        .await
        .expect("clean_table: deleting events");

    sqlx::query(&query_vacuum)
        .execute(db)
        .await
        .expect("clean_table: vacuuming table");
}

async fn insert_normalized(db: &DB) {
    const QUERY: &str = "INSERT INTO normalized (id, type, timestamp, received_at, payload)
        VALUES ($1, $2, $3, $4, $5)";
    let stream = stream::iter(0..EXECUTIONS);
    let base_event = generate_event();

    stream
        .for_each_concurrent(CONCURRENCY as usize, |_| {
            let mut event = base_event.clone();
            event.id = Uuid::new_v4();
            async move {
                sqlx::query(QUERY)
                    .bind(&event.id)
                    .bind(&event.r#type)
                    .bind(&event.timestamp)
                    .bind(&event.received_at)
                    .bind(&event.payload)
                    .execute(db)
                    .await
                    .expect("normalized: inserting event");
            }
        })
        .await;
}

async fn insert_key_value(db: &DB) {
    const QUERY: &str = "INSERT INTO key_value (key, value)
        VALUES ($1, $2)";
    let stream = stream::iter(0..EXECUTIONS);
    let base_event = generate_event();
    let key_value_event = KeyValueEvent {
        key: base_event.id,
        value: serde_json::to_vec(&base_event).expect("key_value: serializing event"),
    };

    stream
        .for_each_concurrent(CONCURRENCY as usize, |_| {
            let mut event = key_value_event.clone();
            event.key = Uuid::new_v4();
            async move {
                sqlx::query(QUERY)
                    .bind(&event.key)
                    .bind(&event.value)
                    .execute(db)
                    .await
                    .expect("key_value: inserting event");
            }
        })
        .await;
}

async fn insert_key_value_compressed(db: &DB) {
    const QUERY: &str = "INSERT INTO key_value_compressed (key, value)
        VALUES ($1, $2)";
    let stream = stream::iter(0..EXECUTIONS);
    let base_event = generate_event();
    let key_value_event = KeyValueEvent {
        key: base_event.id,
        value: serde_json::to_vec(&base_event).expect("key_value_compressed: serializing event"),
    };

    stream
        .for_each_concurrent(CONCURRENCY as usize, |_| {
            let mut event = key_value_event.clone();
            event.key = Uuid::new_v4();
            async move {
                event.value = tokio::task::spawn_blocking(move || {
                    zstd::bulk::compress(&event.value, 2)
                        .expect("key_value_compressed :compressing event")
                })
                .await
                .expect("key_value_compressed: awaiting for tokio::spawn_blocking");

                sqlx::query(QUERY)
                    .bind(&event.key)
                    .bind(&event.value)
                    .execute(db)
                    .await
                    .expect("key_value_compressed: inserting event");
            }
        })
        .await;
}
